src/analyses/ai.h

```diff
@@ -162,6 +162,17 @@ class ai_baset
     fixedpoint(goto_function.body, goto_functions, ns);
   }
 
+  /// \brief Returns the abstract state before the given instruction
+  virtual const ai_domain_baset & abstract_pre(
```

Comment:
Nit: The `\brief` tag isn't actually required here (just `/// Returns the abstract state before the given instruction` is sufficent). 

You might consider giving this function a more meaningful name `abstract_state_pre_instruction`? `get_state_pre_instruction`? Or perhaps taking an `enum state_when { BEFORE_INSTRUCTION, AFTER_INSTRUCTION }`?

Reply:
Fixed and renamed to abstract_precondition and abstract_postcondition.